node/src/components/block_validator.rs

```diff
@@ -98,18 +112,195 @@ pub(crate) enum Event {
+                    ?current_state,
+                    "trying to switch to SignaturesValidationState::AwaitingFinality"
+                );
+            }
+        }
+    }
+
+    fn block_stored(&mut self, stored_block_height: u64) -> Option<NodeId> {
+        match self {
+            SignaturesValidationState::AwaitingBlockStored {
+                sender,
+                block_height,
+            } => {
+                if stored_block_height > *block_height {
+                    debug!(%stored_block_height, awaiting_block_height = %block_height,
+                        "stored higher block; switching back to Pending");
+                    let sender = *sender;
+                    *self = SignaturesValidationState::Pending;
+                    return Some(sender);
+                }
+            }
+            state => {
+                debug!(
+                    ?state,
+                    "handling block finality while not in AwaitingFinality state"
+                );
+            }
+        }
+        None
+    }
+
+    fn require_signatures(
+        &mut self,
+        missing_sigs: HashSet<FinalitySignatureId>,
+        in_flight: KeyedCounter<FinalitySignatureId>,
+    ) {
+        match self {
+            SignaturesValidationState::Pending => {
+                if missing_sigs.is_empty() {
+                    debug!("no signatures required; switching to Valid");
+                    *self = SignaturesValidationState::Valid;
+                } else {
+                    debug!(num_missing = %missing_sigs.len(), "switching to FetchingSignatures");
+                    *self = SignaturesValidationState::FetchingSignatures {
+                        missing_sigs,
+                        in_flight,
+                    };
+                }
+            }
+            state => {
+                warn!(
+                    ?state,
+                    "trying to switch to FetchingSignatures while not in Pending state"
+                );
+            }
+        }
+    }
+
+    fn signature_fetched(&mut self, fetched_sig: FinalitySignatureId) {
+        match self {
+            SignaturesValidationState::FetchingSignatures {
+                missing_sigs,
+                in_flight,
+            } => {
+                missing_sigs.remove(&fetched_sig);
+                in_flight.dec(&fetched_sig);
+                debug!(?fetched_sig, num_missing = %missing_sigs.len(), "fetched a signature");
+                if missing_sigs.is_empty() {
+                    debug!("no more signatures missing, switching state to Valid");
+                    *self = SignaturesValidationState::Valid;
+                }
+            }
+            state => {
+                warn!(
+                    ?state,
+                    "handling fetched signature while not in FetchingSignatures state"
+                );
+            }
+        }
+    }
+
+    fn signature_fetch_failed(&mut self, failed_sig: FinalitySignatureId) {
```

Comment:
If we requested the signature from multiple sources, even if one of them responded with an invalid one, we might still be able to get a valid one from the other sources.

I called it `signature_fetch_failed` because it's called when the fetcher returns an error. A case where the signature was received but was invalid is still a fetch failure in this sense.